accepted/00000-enable-ansible-control-node-in-registered-system.md

```diff
@@ -0,0 +1,136 @@
+* The SSH key to connect the Ansible managed client referenced in the Ansible inventory is used to build the ProxyCommand to use in "salt-ssh".
+
+NOTE: It needs to be clarified in the documentation that `ansible_ssh_private_key_file` needs to be defined in the Ansible inventory of your Ansible control node.
+
+* Once we have this, we can trigger the bootstrap of those systems by calling "salt-ssh" like this:
+
+```console
+# salt-ssh --ssh-option='ProxyCommand="/usr/bin/ssh -i /srv/susemanager/salt/salt_ssh/mgr_ssh_id -o StrictHostKeyChecking=no -o User=root -W %h:%p uyuni-ansible-control-node.tf.local"' --roster=ansible --roster-file=/var/cache/ansible/ansible-inventory.yaml -N webserver1 state.apply certs,bootstrap pillar='{"mgr_server": "uyuni-srv.tf.local", "minion_id": "uyuni-ansible-sles15sp1-2.tf.local"}'
+```
+In esence, we apply the same states that we do at the time of Boostrapping a new minion via the UI, passing the necessary information as pillar data.
+
+* Another example of usage (without using the ansible roster/targetting):
+
+```console
+# salt-ssh --ssh-option='ProxyCommand="/usr/bin/ssh -i PATH_TO_SALT_SSH_KEY_IN_THE_UYUNI_SERVER -o StrictHostKeyChecking=no -W %h:%p uyuni-stable-ansible-controller.tf.local /usr/bin/ssh -i PATH_TO_THE_SSH_KEY_IN_ANSIBLE_CONTROL_NODE uyuni-stable-ansible-opensuse152-2.tf.local" -o StrictHostKeyChecking=no' uyuni-stable-ansible-opensuse152-2.tf.local test.ping
+```
+
+
+### Triggering playbook executions in an Ansible control node
+
+After exploring the Ansible control node, inventory and playbooks, we can allow Uyuni to trigger the execution of one of those playbooks via a simple Salt state to execute in that control node:
+
+```yaml
+execute_ansible_playbook:
+  ansible.playbooks:
+    - name: /some/path/in/my/ansible/control/node/playbook_1.yaml
+```
+
+We would need to create a new type of Action in Uyuni for "running a playbook". We want this action to appear in the system event history for the Ansible control node.
+
+
+### Shipping Ansible for SLE15+
+
+At the time of writing this RFC, Ansible is not shipped in SLE15 any SP. In order to allow a posible Ansible control node running on SLE, we would need to ship Ansible probably in the SLE15 client tools channels.
+
+- Which version to ship? Probably Ansible 2.9, which is currently packaged for Leap, and we have it also in Factory: https://build.opensuse.org/package/show/openSUSE:Factory/ansible
+
+Considerations and other discussions about shipping Ansible in SLE, the support level, etc, can be discussed outside this RFC since it's out of the main focus of the general Ansible integration in Uyuni.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we **not** do this?
+
+  * obscure corner cases
+  * will it impact performance? Bootstrapping Ansible managed clients might take some time since "salt-ssh" is involved.
```

Comment:
Say I have 1,000 Ansible clients and I want to bootstrap them with salt-ssh. Will it take a lot more time than if I bootstrap them with the normal bootstrap? To me, it's the same, the only additional step is getting the SSH key from Ansible. Or am I missing something?

Reply:
Using our UI you can only register 1 and even with API you register one after another.

Here I would expect some kind of parallel onboarding and this requires batches and other things.
Performance is a possible problem. It depends on the implementation.

Reply:
It should take the same time in both cases. In the end, the "Boostrapping UI" page is basically also doing "salt-ssh" execution under the hood.

As you mentioned, the only additional step would be to fetch the SSH keys from the Ansible control node first, and not necessary for each client to bootstrap but just once at the time of getting the Ansible inventory.